### PR TITLE
Feature Extractor: Wav2Vec2 & Speech2Text - Allow truncation + padding=longest

### PR DESCRIPTION
**Pull request by patrickvonplaten**
# What does this PR do?

For training is speech it's quite important that we are able to make use of the `padding="longest"` *a.k.a* dynamic batching in PyTorch on GPU since speech inputs can vary from 2 seconds to up to 30 seconds. We also want to be able to use truncation with `padding="longest"` in such a case. This PR enables using truncation (*i.e.* cutting sequences to a max length of x seconds) with the `padding="longest"` mode.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: src/transformers/feature_extraction_sequence_utils.py [hunk 1]
```diff
@@ -322,9 +323,7 @@ def _truncate(
         if not truncation:
             return processed_features
         elif truncation and max_length is None:
-            raise ValueError(
-                "When setting ``truncation=True``, make sure that ``max_length`` is defined and ``padding='max_length'``"
-            )
+            raise ValueError("When setting ``truncation=True``, make sure that ``max_length`` is defined.")
```
Comment (patrickvonplaten) [inline src/transformers/feature_extraction_sequence_utils.py:326]: Now we do allow truncation=True + padding="longest"

File: src/transformers/models/wav2vec2/feature_extraction_wav2vec2.py [hunk 2]
```diff
@@ -91,7 +91,7 @@ def zero_mean_unit_var_norm(
 
             for vector, length in zip(input_values, attention_mask.sum(-1)):
                 normed_slice = (vector - vector[:length].mean()) / np.sqrt(vector[:length].var() + 1e-7)
-                if length > normed_slice.shape[0]:
+                if length < normed_slice.shape[0]:
```
Comment (patrickvonplaten) [inline src/transformers/models/wav2vec2/feature_extraction_wav2vec2.py:94]: added test so that such a bug is caught next time

File: src/transformers/feature_extraction_sequence_utils.py [hunk 3]
```diff
@@ -211,16 +211,17 @@ def pad(
...
                 max_length=max_length,
                 pad_to_multiple_of=pad_to_multiple_of,
                 truncation=truncation,
             )
-            truncated_inputs.append(inputs)
+            truncated_inputs.append(inputs_slice)
 
         if padding_strategy == PaddingStrategy.LONGEST:
-            max_length = max(len(inputs) for inputs in required_input)
+            # make sure that `max_length` cannot be longer than longest truncated length
```
Comment (sgugger) [inline src/transformers/feature_extraction_sequence_utils.py:223]:
```suggestion
            # make sure that `max_length` cannot be longer than the longest truncated length
```

**Review by sgugger — APPROVED** [review]
Thanks for adding this and nice new tests!

**Review by anton-l — APPROVED** [review]
LGTM, thanks for catching the bug :)

File: tests/test_feature_extraction_speech_to_text.py [hunk 4]
```diff
@@ -214,3 +216,49 @@ def _check_zero_mean_unit_variance(input_vector):
         _check_zero_mean_unit_variance(input_features[0, : fbank_feat_lengths[0]])
         _check_zero_mean_unit_variance(input_features[1])
         _check_zero_mean_unit_variance(input_features[2])
+
+    def test_cepstral_mean_and_variance_normalization_trunc_longest(self):
```
Comment (LysandreJik) [inline tests/test_feature_extraction_speech_to_text.py:220]: Very nice test

**Review by LysandreJik — APPROVED** [review]
LGTM
